library/core/src/array/mod.rs

```diff
@@ -426,6 +427,59 @@ impl<T: Clone, const N: usize> Clone for [T; N] {
     }
 }
 
+#[unstable(feature = "random", issue = "130703")]
+impl<T: Random, const N: usize> Random for [T; N] {
+    fn random(source: &mut (impl RandomSource + ?Sized)) -> Self {
+        SpecArrayRandom::random(source)
+    }
+}
+
+#[unstable(feature = "random", issue = "130703")]
+trait SpecArrayRandom: Sized {
+    fn random<const N: usize>(source: &mut (impl RandomSource + ?Sized)) -> [Self; N];
+}
+
+#[unstable(feature = "random", issue = "130703")]
+impl<T: Random> SpecArrayRandom for T {
+    default fn random<const N: usize>(source: &mut (impl RandomSource + ?Sized)) -> [T; N] {
+        from_fn(|_| T::random(source))
+    }
+}
+
+macro_rules! impl_random_for_integer_array {
+    ($t:ty) => {
+        #[unstable(feature = "random", issue = "130703")]
+        impl SpecArrayRandom for $t {
+            fn random<const N: usize>(source: &mut (impl RandomSource + ?Sized)) -> [$t; N] {
+                let mut buf = [const { MaybeUninit::<$t>::uninit() }; N];
+                // SAFETY: all elements in the buffer were initialized with
+                // random bytes.
+                unsafe {
+                    let bytes = slice::from_raw_parts_mut(
+                        &raw mut buf as *mut u8,
+                        N * (<$t>::BITS as usize / 8),
+                    );
+                    source.fill_bytes(bytes);
```

Comment:
This is unsound: you create a `&mut [u8]` and pass it to user code – but nothing guarantees that the user code won't try to read the slice, which would be undefined behaviour. My recommendation would be to use `MaybeUninit::zeroed` to create `buf`.

Reply:
Ideally this needs a trait bound like [`zerocopy::FromBytes`](https://docs.rs/zerocopy/latest/zerocopy/trait.FromBytes.html) or [`bytemuck::Pod`](https://docs.rs/bytemuck/latest/bytemuck/trait.Pod.html), but we don't have that in `std`, so... the macro should be renamed to `unsafe_impl_...` since safety depends on the macro caller?